bench/macro/lsm-tree-bench-wp8.hs

```diff
@@ -0,0 +1,404 @@
+
+-------------------------------------------------------------------------------
+-- clock
+-------------------------------------------------------------------------------
+
+timed :: IO a -> IO (a, Double)
+timed action = do
+    t1 <- Clock.getTime Clock.Monotonic
+    x  <- action
+    t2 <- Clock.getTime Clock.Monotonic
+    let !t = fromIntegral (Clock.toNanoSecs (Clock.diffTimeSpec t2 t1)) * 1e-9
+    return (x, t)
+
+timed_ :: IO () -> IO Double
+timed_ action = do
+    t1 <- Clock.getTime Clock.Monotonic
+    action
+    t2 <- Clock.getTime Clock.Monotonic
+    return $! fromIntegral (Clock.toNanoSecs (Clock.diffTimeSpec t2 t1)) * 1e-9
+
+-------------------------------------------------------------------------------
+-- setup
+-------------------------------------------------------------------------------
+
+-- https://input-output-hk.github.io/fs-sim
+doSetup :: GlobalOpts -> SetupOpts -> IO ()
+doSetup gopts _opts = do
+    let mountPoint :: FS.MountPoint
+        mountPoint = FS.MountPoint gopts.rootDir
+
+    let someFs :: FS.SomeHasFS IO
+        someFs = FS.SomeHasFS (FsIO.ioHasFS mountPoint)
+
+    name <- maybe (fail "invalid snapshot name") return $
+        LSM.mkSnapshotName "bench"
+
+    withSession someFs (FS.mkFsPath []) $ \session -> do
+        tbh <- LSM.new @IO @K @V @B session defaultTableConfig
+
+        forM_ [ 0 .. gopts.initialSize ] $ \ (fromIntegral -> i) -> do
+            -- TODO: this procedure simply inserts all the keys into initial lsm tree
+            -- We might want to do deletes, so there would be delete-insert pairs
+            -- Let's do that when we can actually test that benchmark works.
+
+            let k = makeKey i
+            let v = theValue
+
+            -- TODO: LSM.inserts has annoying order
+            flip LSM.inserts tbh
+                [ (k, v, Nothing)
+                ]
+
+        LSM.snapshot name tbh
+
+-------------------------------------------------------------------------------
+-- dry-run
+-------------------------------------------------------------------------------
+
+doDryRun :: GlobalOpts -> RunOpts -> IO ()
+doDryRun gopts opts = do
+    time <- timed_ $ doDryRun' gopts opts
+    printf "Batch generation: %.03f sec\n" time
+
+doDryRun' :: GlobalOpts -> RunOpts -> IO ()
+doDryRun' gopts opts = do
+    -- calculated some expected statistics for generated batches
+    id $ do
+        -- we generate n random numbers in range of [ 1 .. d ]
+        -- what is the chance they are all distinct
+        let n = fromIntegral (opts.batchCount * opts.batchSize) :: Double
+        let d = fromIntegral gopts.initialSize :: Double
+        -- this is birthday problem.
+        let p = 1 - exp (negate $  (n * (n - 1)) / (2 * d))
+
+        -- number of people with a shared birthday
+        -- https://en.wikipedia.org/wiki/Birthday_problem#Number_of_people_with_a_shared_birthday
+        let q = n * (1 - ((d - 1) / d) ** (n - 1))
```

Comment:
Very nice that you included the calculations and also the actual check!

One detail: The check below counts the collisions, but this formula calculates the number of people with shared birthday, i.e. 2 if one collision exists, which is about a factor 2 more. So both make sense, but since we look at them next to each other, it would be good for them to agree.

Reply:
Once this one is addressed (which could be telling me why I'm wrong) the PR seems good to merge.

Reply:
> number of people with a shared birthday

is an overestimation, if there are three people with same birthdays it's only two collisions etc.

I can look out for formula for collisions, but not right now. We might as well scrap this way of key generation too.

I.e. if you don't like this formula, then I'd rather remove whole calculation as the quick solution.

Reply:
Ah no, don't remove it. It's nice to have.

Okay, then there's a very quick solution: Just add a comment like "this overestimates the number of collisions by approximately factor of 2 (assuming it's extremely unlikely for a key to occur more than twice)". I think it's good to have that mentioned in case someone wonders about the the mismatch in the output.

I also think it would then be nicer to print `q / 2` instead of `q`, but you can leave it as is.

Reply:
done